### PR TITLE
Feat: Add shared memory functionality and tests

### PR DESCRIPTION
**Pull request by noluyorAbi**
### **User description**
# Pull Request: Implementierung des Shared-Memory-Bereichs

## Beschreibung

Dieser Pull Request fügt dem Projekt einen **Shared-Memory-Bereich** hinzu, der die Kommunikation zwischen den Prozessen **Gameserver**, **Thinker** und **Connector** optimiert. Die Implementierung unterstützt eine dynamische Anzahl von Spielern und gewährleistet eine effiziente Speicherverwaltung.

## Änderungen

1. **Shared-Memory-Module hinzugefügt:**
   - **Dateien:**
     - `modules/shared_memory/shared_memory.c`
     - `modules/shared_memory/shared_memory.h`
   - **Strukturen:**
     ```c
     typedef struct {
       char gameName[256];     // Spielname vom Gameserver
       int playerNumber;       // Eigene Spielernummer
       int playerCount;        // Anzahl der Spieler
       pid_t thinkerPID;       // PID des Thinker-Prozesses
       pid_t connectorPID;     // PID des Connector-Prozesses
       Player players[];       // Dynamisches Array der Spieler
     } SharedMemory;

     typedef struct {
       int playerNumber;       // Spielernummer
       char playerName[128];   // Name des Spielers
       bool isRegistered;      // Registrierungsstatus
     } Player;
     ```
   - **Funktionen:**
     - `createSharedMemory(int numPlayers)`: Reserviert ein Shared-Memory-Segment.
     - `attachSharedMemory(int shmid)`: Bindet das Segment an den Adressraum.
     - `detachSharedMemory(SharedMemory *shm)`: Trennt das Segment.
     - `removeSharedMemory(int shmid)`: Entfernt das Segment.

2. **Integration in das Hauptprojekt:**
   - **`main.c`:**
     - Einbindung von `shared_memory.c`.
     - Nutzung der Shared-Memory-Datenstruktur zur Verwaltung der Spielparameter.
   - **`args_parser.c` & `args_parser.h`:**
     - Verbesserte Argumentenparsing-Logik zur Unterstützung dynamischer Spielerzahlen.

3. **Tests hinzugefügt:**
   - **Dateien:**
     - `tests/test_sharedMemory/test_sharedMemory.c`
     - `tests/test_sharedMemory/README.md`
   - **Testfälle:**
     - Initialisierung und Anbindung von Shared-Memory für verschiedene Spielerzahlen.
     - Datenmanipulation und Validierung der Spielerinformationen.
     - Fehlerbehandlung bei ungültigen Shared-Memory-IDs und -Zeigern.
     - Sicherstellung der korrekten Trennung und Entfernung der Shared-Memory-Segmente.

4. **CI-Pipeline angepasst (`.github/workflows/ci.yml`):**
   - **Workflow-Name geändert:**
     - Von `C Project Build, Memory Leak Check, and Deployment Pipeline` zu `C Project Build, Memory Leak Check, and Shared Memory Test`.
   - **Neue Schritte hinzugefügt:**
     - Kompilierung und Ausführung der Shared-Memory-Tests.
     - Valgrind-Analyse zur Überprüfung auf Speicherlecks.
   - **Deployment-Schritt entfernt:** Fokus auf Build und Test.

5. **Projektstruktur und Dokumentation aktualisiert:**
   - **`.gitignore`:**
     - Hinzufügen des `/bin`-Verzeichnisses:
       ```
       # Bin directory
       /bin
       ```
   - **`README.md`:**
     - Aktualisierung der Verzeichnisstruktur.
     - Beschreibung der neuen Shared-Memory-Module und der zugehörigen Tests.

## Begründung

- **Effiziente Kommunikation:** Shared Memory ermöglicht eine schnelle und ressourcenschonende Kommunikation zwischen Prozessen, ideal für Multiplayer-Anwendungen.
- **Skalierbarkeit:** Die flexible Datenstruktur unterstützt zukünftige Erweiterungen und verschiedene Spieltypen mit unterschiedlicher Spieleranzahl.
- **Ressourcenmanagement:** Sicherstellung der ordnungsgemäßen Freigabe von Shared-Memory-Segmenten verhindert Speicherlecks.
- **Automatisierte Qualitätssicherung:** Integration der Tests in die CI-Pipeline gewährleistet die Stabilität und Fehlerfreiheit der neuen Funktionalitäten.

## Zusammenfassung

Mit dieser Implementierung wird die Speicherverwaltung des Projekts optimiert und die Grundlage für eine skalierbare Multiplayer-Architektur geschaffen. Die umfassenden Tests und Anpassungen in der CI-Pipeline sichern die Zuverlässigkeit und Qualität der neuen Shared-Memory-Funktionalität.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added a new shared memory module (`shared_memory.c` and `shared_memory.h`) to manage inter-process communication efficiently.
- Integrated shared memory functionality into the main project (`main.c`).
- Implemented tests for shared memory operations, including initialization, error handling, and boundary conditions.
- Updated CI workflow to include shared memory tests and Valgrind analysis for memory leak detection.
- Improved documentation in `args_parser` module and added detailed documentation for shared memory tests.
- Updated project structure and README to reflect the addition of the shared memory module and tests.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>main.c</strong><dd><code>Integrate shared memory module and refactor constants usage</code></dd></summary>
<hr>

main.c

<li>Integrated the <code>shared_memory</code> module into the main project.<br> <li> Replaced direct usage of <code>config</code> fields with constants for <code>GAME_ID</code> and <br><code>PLAYER_NUMBER</code>.<br> <li> Improved readability of <code>printf</code> statements.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/21/files#diff-a0cb465674c1b01a07d361f25a0ef2b0214b7dfe9412b7777f89add956da10ec">+7/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>shared_memory.c</strong><dd><code>Implement shared memory operations with error handling</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

modules/shared_memory/shared_memory.c

<li>Implemented shared memory functions: creation, attachment, detachment, <br>and removal.<br> <li> Added error handling for shared memory operations.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/21/files#diff-ef626cb2231e9da56a96c0ab9a135d18122bdd03a773f2189a97f1fb2a0b1362">+43/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>shared_memory.h</strong><dd><code>Define structures and headers for shared memory</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

modules/shared_memory/shared_memory.h

<li>Defined <code>Player</code> and <code>SharedMemory</code> structures for shared memory usage.<br> <li> Included necessary headers for shared memory operations.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/21/files#diff-fa7ff423c34e39bf6d340aced3be463044107164b424cf0cd760519301252b77">+24/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>args_parser.c</strong><dd><code>Add comments and improve argument parsing logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

modules/args_parser/args_parser.c

<li>Added comments to improve code readability.<br> <li> Enhanced parsing logic for command-line arguments.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/21/files#diff-f727819942a376e270f04716af07fa1ae3baebe386b33fba08892e84ffb3dd96">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>args_parser.h</strong><dd><code>Simplify and translate comments in header file</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

modules/args_parser/args_parser.h

<li>Simplified and translated comments to English.<br> <li> Improved documentation for <code>GameConfig</code> structure and functions.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/21/files#diff-8dfd9680ddc61db50f0b7df4b80b17772ac2ff965dd19f1b290f1f529b7bbc6e">+6/-19</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Update README with shared memory module details</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

README.md

<li>Updated project structure to include shared memory module.<br> <li> Documented new test directory for shared memory.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/21/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5">+12/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Add documentation for shared memory tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_sharedMemory/README.md

<li>Documented the purpose and functionality of shared memory tests.<br> <li> Provided detailed test scenarios and expected results.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/21/files#diff-8d44738906a755521d464db2982f979eb19df2412cd075a75413a23f9391f88a">+91/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>test_sharedMemory.c</strong><dd><code>Add integration tests for shared memory operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_sharedMemory/test_sharedMemory.c

<li>Added integration tests for shared memory operations.<br> <li> Tested initialization, error handling, and boundary conditions.<br> <li> Validated memory management using Valgrind.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/21/files#diff-9ba6e22f6c10f159f5ca8a19ff956df04bd21ee457c9650dfc061039d9ecbc93">+103/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>ci.yml</strong><dd><code>Update CI workflow to include shared memory tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/ci.yml

<li>Updated CI workflow to include shared memory tests.<br> <li> Added Valgrind analysis for memory leak detection.<br> <li> Improved workflow comments for clarity.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/21/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03f">+39/-9</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Review by jpeter96 — DISMISSED** [review]

File: modules/shared_memory/shared_memory.c [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  if (shmaddr == (void *) -1) {
+    fprintf(stderr, "shmat failed: %s\n", strerror(errno));
+    return NULL;
+  }
+  return (SharedMemory *) shmaddr;
+}
+
+int detachSharedMemory(SharedMemory *shm) {
+  if (shmdt((void *) shm) == -1) {
+    fprintf(stderr, "shmdt failed: %s\n", strerror(errno));
```
Comment (mvritz) [inline modules/shared_memory/shared_memory.c:31]:
```suggestion
fprintf(stderr, "[ERROR] Failed to create shared memory: %s\n", strerror(errno));
```

File: modules/shared_memory/shared_memory.c [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+SharedMemory *attachSharedMemory(int shmid) {
+  void *shmaddr = shmat(shmid, NULL, 0);
+  if (shmaddr == (void *) -1) {
+    fprintf(stderr, "shmat failed: %s\n", strerror(errno));
+    return NULL;
+  }
+  return (SharedMemory *) shmaddr;
+}
+
+int detachSharedMemory(SharedMemory *shm) {
```
Comment (mvritz) [inline modules/shared_memory/shared_memory.c:29]: docs?

Comment (mvritz) [inline modules/shared_memory/shared_memory.c:29]: I would add something like this tbh as an utility function:

size_t calculateShmSize(int numPlayers) {
    return sizeof(SharedMemory) + numPlayers * sizeof(Player);
}

**Review by mvritz — CHANGES_REQUESTED** [review]

**Comment by noluyorAbi** [comment]
# Detaillierte Änderungen am Shared Memory Modul

## Übersicht

In diesem Commit wurden umfassende Verbesserungen und Erweiterungen am Shared Memory Modul vorgenommen. Die Änderungen umfassen:

- **Verbesserte Fehlerbehandlung** in den Shared Memory Funktionen.
- **Erweiterung der Funktionalität** zur Initialisierung und Bereinigung des Shared Memory Segments.
- **Refaktorierung der Header-Datei** zur besseren Strukturierung und Dokumentation.
- **Erstellung und Aktualisierung der Testfälle** zur Sicherstellung der Funktionalität und Robustheit.
- **Entfernung veralteter Testdateien**.

Im Folgenden werden die Änderungen im Detail erläutert.

## Änderungen in `shared_memory.c`

### 1. Header-Includes

- **Änderung des Include-Pfads:**
  ```diff
  -#include "shared_memory.h"
  +#include "./shared_memory.h"
  ```
  Der Include-Pfad wurde explizit auf das aktuelle Verzeichnis gesetzt, um mögliche Konflikte mit anderen Header-Dateien zu vermeiden.

- **Zusätzliche Includes:**
  ```diff
  +#include <unistd.h>
  ```
  Die Bibliothek `<unistd.h>` wurde hinzugefügt, um Funktionen wie `getpid()` nutzen zu können.

### 2. Definitionen und Konstanten

- **Definition der maximalen Spielerzahl:**
  ```c
  #define MAX_PLAYERS 100 // Definieren Sie eine maximale Anzahl von Spielern
  ```
  Eine Konstante zur Begrenzung der Anzahl der Spieler wurde hinzugefügt, um Überläufe und unerwartetes Verhalten zu verhindern. **Hinweis:** `MAX_PLAYERS` ist derzeit auf 100 gesetzt, um umfangreiche Testfälle zu ermöglichen. In der Produktionsumgebung wird die Anzahl der Spieler jedoch durch den `args_parser` auf 1 oder 2 begrenzt, sodass `MAX_PLAYERS` für Testspezifische Zwecke auf 100 belassen werden kann.

### 3. Verbesserte Fehlerbehandlung

- **Validierung der Eingabeparameter:**
  ```c
  if (numPlayers < 0 || numPlayers > MAX_PLAYERS) {
    fprintf(stderr, "Ungültige Anzahl von Spielern: %d. Muss zwischen 0 und %d liegen.\n", numPlayers, MAX_PLAYERS);
    return SHM_ERROR_CREATION;
  }
  ```
  Eingaben werden nun validiert, um sicherzustellen, dass die Anzahl der Spieler innerhalb der erlaubten Grenzen liegt. Zusätzlich übernimmt der `args_parser` bereits die Überprüfung, ob die Anzahl der Spieler auf 1 oder 2 beschränkt ist, wodurch in der Regel `numPlayers` nicht über diese Werte hinausgehen sollte.

- **Verwendung von `perror` für Fehlermeldungen:**
  ```diff
  -fprintf(stderr, "shmget failed: %s\n", strerror(errno));
  +perror("shmget fehlgeschlagen");
  ```
  `perror` wird verwendet, um detaillierte Fehlermeldungen basierend auf `errno` auszugeben.

### 4. Rückgabewerte und Fehlercodes

- **Einführung von spezifischen Fehlercodes:**
  ```c
  typedef enum {
    SHM_SUCCESS = 0,
    SHM_ERROR_CREATION = -1,
    SHM_ERROR_ATTACH = -2,
    SHM_ERROR_DETACH = -3,
    SHM_ERROR_REMOVE = -4,
  } ShmError;
  ```
  Die Funktionen geben nun spezifische Fehlercodes zurück, anstatt nur `-1`, um eine genauere Fehlerdiagnose zu ermöglichen.

### 5. Erweiterte Funktionalitäten

- **Initialisierung des Shared Memory Segments:**
  ```c
  int initSharedMemory(int numPlayers, const char *gameName, int playerNumber,
                       pid_t thinkerPID, pid_t connectorPID) {
    // Validierungen und Initialisierungen
  }
  ```
  Eine neue Funktion zur Initialisierung des Shared Memory Segments wurde hinzugefügt, die zusätzliche Parameter wie `gameName`, `playerNumber`, `thinkerPID` und `connectorPID` akzeptiert.

- **Bereinigung des Shared Memory Segments:**
  ```c
  void cleanupSharedMemory(int shmid, SharedMemory *shm) {
    // Detach und Remove mit Fehlerüberprüfung
  }
  ```
  Eine Funktion zur sauberen Freigabe des Shared Memory Segments wurde hinzugefügt.

## Änderungen in `shared_memory.h`

### 1. Strukturverbesserungen

- **Erweiterung der `SharedMemory` Struktur:**
  ```c
  typedef struct {
    char gameName[256];
    int playerNumber;
    int playerCount;
    pid_t thinkerPID;
    pid_t connectorPID;
    Player players[];
  } SharedMemory;
  ```
  Zusätzliche Felder wie `gameName`, `thinkerPID` und `connectorPID` wurden hinzugefügt, um mehr Kontextinformationen im Shared Memory zu speichern.

### 2. Fehlercodes und Dokumentation

- **Einführung von Fehlercodes:**
  ```c
  typedef enum {
    SHM_SUCCESS = 0,
    SHM_ERROR_CREATION = -1,
    SHM_ERROR_ATTACH = -2,
    SHM_ERROR_DETACH = -3,
    SHM_ERROR_REMOVE = -4,
  } ShmError;
  ```
  Spezifische Fehlercodes wurden definiert, um die Fehlerbehandlung zu verbessern.

- **Detaillierte Funktionsdokumentation:**
  Jede Funktion wurde mit ausführlichen Kommentaren versehen, die den Zweck, die Parameter und die Rückgabewerte erläutern. Beispiel:
  ```c
  /**
   * @brief Erstellt ein Shared-Memory-Segment.
   *
   * @param numPlayers Die Anzahl der Spieler, für die Platz reserviert werden soll.
   * @return int Die ID des erstellten Shared-Memory-Segments oder ein Fehlercode (siehe ShmError).
   */
  int createSharedMemory(int numPlayers);
  ```

## Änderungen in den Testdateien (`test_sharedMemory.c`)

### 1. Refaktorierung der Tests

- **Direktes Einbinden des Headers:**
  ```diff
  -#include "../../modules/shared_memory/shared_memory.c"
  +#include "../../modules/shared_memory/shared_memory.h" // Include shared_memory.h directly
  ```
  Die Implementierungsdatei wird nicht mehr direkt eingebunden, stattdessen wird der Header genutzt, um die Funktionen zu testen.

- **Einführung von Testzählern:**
  ```c
  int PASS_COUNT = 0;
  int FAIL_COUNT = 0;
  ```
  Zähler zur Nachverfolgung der bestandenen und fehlgeschlagenen Tests wurden hinzugefügt.

### 2. Erweiterte Testfunktionen

- **Hilfsfunktion zur Berichterstattung:**
  ```c
  void report_test(const char *test_name, bool passed, const char *message) {
    if (passed) {
      printf("[PASS] %s: %s\n", test_name, message);
      PASS_COUNT++;
    } else {
      printf("[FAIL] %s: %s\n", test_name, message);
      FAIL_COUNT++;
    }
  }
  ```
  Eine Funktion zur einheitlichen Darstellung der Testergebnisse wurde eingeführt.

- **Neue Testfälle:**
  - **`testSharedMemoryCreateAndAttach`**: Testet das Erstellen und Anhängen eines Shared Memory Segments.
  - **`testInitSharedMemory`**: Überprüft die Initialisierung des Shared Memory Segments mit verschiedenen Parametern.
  - **`testInvalidInput`**: Testet das Verhalten der Funktionen bei ungültigen Eingaben.
  - **`testErrorHandling`**: Überprüft die Fehlerbehandlung bei ungültigen Operationen.
  - **`testBoundaryConditions`**: Testet die Funktionen an den Grenzwerten der erlaubten Eingaben (z.B. 0 und 100 Spieler).

- **Entfernung veralteter Testdateien:**
  ```diff
  deleted file mode 100755
  index 136c12f..0000000
  Binary files a/tests/test_sharedMemory/test_sharedMemory and /dev/null differ
  ```
  Alte ausführbare Testdateien wurden entfernt, um Platz für die neuen, verbesserten Testskripte zu schaffen.

### 3. Verbesserte Testausgabe

- **Detaillierte Informationsausgaben:**
  ```c
  printf("[INFO] Shared memory created with shmid: %d\n", shmid);
  ```
  Zusätzliche Informationen während der Tests erleichtern das Debugging und die Nachverfolgung des Testverlaufs.

- **Zusammenfassung der Testergebnisse:**
  ```c
  printf("[SUMMARY] Bestanden: %d, Fehlgeschlagen: %d\n", PASS_COUNT, FAIL_COUNT);
  return (FAIL_COUNT == 0) ? 0 : 1;
  ```
  Am Ende der Testausführung wird eine Zusammenfassung der bestandenen und fehlgeschlagenen Tests angezeigt.

## Gründe für die Änderungen

1. **Verbesserte Robustheit:** Durch die Einführung spezifischer Fehlercodes und erweiterter Fehlerbehandlungen wird das Modul robuster und weniger anfällig für unerwartete Eingaben oder Zustände.

2. **Erhöhte Funktionalität:** Die neuen Funktionen zur Initialisierung und Bereinigung des Shared Memory Segments bieten eine bessere Kontrolle und Flexibilität bei der Nutzung des Moduls.

3. **Bessere Testabdeckung:** Die erweiterten und neuen Testfälle stellen sicher, dass das Modul in verschiedenen Szenarien korrekt funktioniert, einschließlich Grenzwert- und Fehlersituationen.

4. **Verbesserte Wartbarkeit:** Durch die Refaktorierung des Codes und die verbesserte Dokumentation in den Header-Dateien wird die Wartbarkeit und Erweiterbarkeit des Moduls verbessert.

5. **Konsistente und verständliche Fehlermeldungen:** Die Verwendung von `perror` und detaillierten Fehlermeldungen erleichtert die Fehlersuche und das Debugging erheblich.

6. **Testzwecke für `MAX_PLAYERS`:** Die Festlegung von `MAX_PLAYERS` auf 100 dient ausschließlich Testzwecken, um umfangreiche Szenarien zu simulieren. In der tatsächlichen Anwendung wird die Anzahl der Spieler jedoch durch den `args_parser` auf 1 oder 2 beschränkt, wodurch die hohe Obergrenze für die Produktion nicht relevant ist.

## Zusammenfassung

Dieser Commit stellt eine signifikante Verbesserung des Shared Memory Moduls dar, indem er die Fehlerbehandlung verbessert, die Funktionalität erweitert, die Testabdeckung erhöht und die Wartbarkeit des Codes verbessert. Diese Änderungen tragen dazu bei, die Zuverlässigkeit und Stabilität des Systems zu gewährleisten und bieten eine solide Grundlage für zukünftige Erweiterungen und Anpassungen. **Zusätzlich wurde klargestellt, dass `MAX_PLAYERS` auf 100 gesetzt ist, um umfangreiche Tests zu ermöglichen, während der `args_parser` bereits sicherstellt, dass nur 1 oder 2 Spieler zugelassen werden.**
